src/E1_coldflow/E1_coldflow.cpp

```diff
@@ -105,26 +100,25 @@ void loop() {
     int i = 0;
     while (RFSerial.available()) {
       command[i] = RFSerial.read();
-      Serial.print(command[i]);
       i++;
     }
+    Serial.write(command, i);
 
     debug(String(command), DEBUG);
-    int action = decode_received_packet(String(command), &valve, valves, numValves);
+    int action = decode_received_packet(command, &valve, valves, numValves, DEBUG);
     if (action != -1) {
       take_action(&valve, action);
-      packet = make_packet(valve.id, false);
-      Serial.println(packet);
+      packet_size = make_packet(valve.id, false);
+      Serial.write(packet, packet_size);
       #if SERIAL_INPUT != 1
-        RFSerial.println(packet);
+        RFSerial.write(packet, packet_size);
       #endif
-      write_to_SD(packet.c_str(), file_name);
+      packet[packet_size] = '\0';
+      write_to_SD(packet, file_name);
```

Comment:
You don't need to specify how many bytes to write from `packet`? What if `packet` has junk left over at the end of it from a previous packet?

Reply:
See above. it looks for the `\0` character